[IR] Add per-function numbers to basic blocks

aengelke opened the pull request:
Every basic block that is linked into a function now has a unique number, which can be queried using getNumber(). Numbers are densely allocated, but not re-assigned on block removal for stability. Block numbers are intended to be fairly stable and only be updated when removing a several basic blocks to make sure the numbering doesn't become too sparse.

To reduce holes in the numbering, renumberBlocks() can be called to re-assign numbers in block order.

Previous discussion in: https://discourse.llvm.org/t/rfc-add-auxiliary-field-for-per-pass-custom-data-to-basicblock/80229

---

I added a validate method to catch cases where something goes wrong, even if I can't really imagine how invalid numbers can occur. But I think it's better to be safe and rule out this potential source of bugs when more things depend on the numbering.

Not sure if I should add a `getNumBlockIDs()` that returns NextBlockNum. So far, I haven't felt the need. I think we can always add it later if it turns out to be beneficial.

aengelke committed:
[IR] Add per-function numbers to basic blocks

Every basic block that is linked into a function now has a unique
number, which can be queried using getNumber(). Numbers are densely
allocated, but not re-assigned on block removal for stability. Block
numbers are intended to be fairly stable and only be updated when
removing a several basic blocks to make sure the numbering doesn't
become too sparse.

To reduce holes in the numbering, renumberBlocks() can be called to
re-assign numbers in block order.

nikic commented on llvm/include/llvm/IR/Function.h:
```diff
@@ -996,6 +1012,12 @@ class LLVM_EXTERNAL_VISIBILITY Function : public GlobalObject,
+/// In release builds, this is a no-op. For !NDEBUG builds, the checks are
+/// implemented in the .cpp file.
+inline void Function::validateBlockNumbers() const {}
+#endif
```
I'd prefer not to have the conditional definitions. Especially as the only callers are currently in the same file, there is no penalty to having the ifdef NDEBUG over the function body only.

aengelke replied:
Agreed, changed

nikic commented on llvm/include/llvm/IR/Function.h:
```diff
@@ -75,6 +75,22 @@ class LLVM_EXTERNAL_VISIBILITY Function : public GlobalObject,
+
+  /// Assert that all blocks have unique numbers within 0..NextBlockNum. This
+  /// has O(n) runtime complexity.
+  void validateBlockNumbers() const;
```
Can make this private, or do you expect this to be called from more places in future patches?

nikic commented on llvm/unittests/IR/FunctionTest.cpp:
```diff
@@ -487,6 +487,78 @@ TEST(FunctionTest, EraseBBs) {
 
+TEST(FunctionTest, BasicBlockNumbers) {
+  LLVMContext Context;
+  Type *VoidType(Type::getVoidTy(Context));
+  FunctionType *FuncType(FunctionType::get(VoidType, false));
```
```suggestion
  Type *VoidType = Type::getVoidTy(Context);
  FunctionType *FuncType = FunctionType::get(VoidType, false);
```

aengelke committed:
Address comments

aengelke committed:
Add getMaxBlockNumber() and getBlockNumberEpoch()

aengelke commented:
Changed. I also decided to add getMaxBlockNumber() here. I think this will come useful.

Furthermore, I introduced the concept of "block number epoch" -- an integer that changes for every renumbering. This will be useful (=already was for me) for identifying use of block numbers after renumbering: on initialization, the current epoch is stored, and on all subsequent accesses, equality with the current epoch can be asserted.

nikic commented on llvm/include/llvm/IR/Function.h:
```diff
@@ -75,6 +75,38 @@ class LLVM_EXTERNAL_VISIBILITY Function : public GlobalObject,
+  /// Epoch of block numbers. (Could be shrinked to uint8_t if required.)
+  unsigned BlockNumEpoch = 0;
+
+public:
```
Can you please move these functions after all the variable declarations? It makes it hard to see what the layout of the structure is if everything is interleaved.

nikic reviewed (approved):
LGTM

aengelke committed:
Move block number functions to block-related functions
